File: scgateway/ios/Classes/ScgatewayEventsHandler.swift
import Foundation
import Flutter
import SCGateway

class ScgatewayEventsHandler: NSObject {
    
    private var eventChannel: FlutterEventChannel?
    private var eventStreamHandler: ScgatewayEventsStreamHandler?
    
    func setup(with registrar: FlutterPluginRegistrar) {
        print("Setting up SCGateway events channel...")
        
        eventStreamHandler = ScgatewayEventsStreamHandler()
        eventChannel = FlutterEventChannel(name: "scgateway_events", binaryMessenger: registrar.messenger())
        eventChannel?.setStreamHandler(eventStreamHandler)
        
        setupNotificationListener()
        print("SCGateway events channel setup complete")
    }
    
    private func setupNotificationListener() {
        NotificationCenter.default.addObserver(
            forName: SCGateway.scgNotificationName,
            object: nil,
            queue: .main
        ) { [weak self] notification in
            self?.handleNotification(notification)
        }
    }
    
    private func handleNotification(_ notification: Notification) {
        guard let eventStreamHandler = eventStreamHandler,
              let userInfo = notification.userInfo,
              let jsonString = userInfo[SCGNotification.strigifiedPayloadKey] as? String else { return }
        
        eventStreamHandler.sendEvent(jsonString)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
        eventChannel?.setStreamHandler(nil)
    }
}

File: scgateway/android/src/main/kotlin/com/example/scgateway_flutter_plugin/ScgatewayEventsHandler.kt
package com.example.scgateway_flutter_plugin

import android.util.Log
import com.smallcase.gateway.data.listeners.Notification
import com.smallcase.gateway.data.listeners.NotificationCenter
import com.smallcase.gateway.portal.ScgNotification
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

class ScgatewayEventsHandler : FlutterPlugin {
    
    private val TAG = "ScgatewayEventsHandler"
    private lateinit var eventChannel: EventChannel
    private lateinit var eventStreamHandler: ScgatewayEventsStreamHandler
    private var notificationObserver: ((Notification) -> Unit)? = null
    
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        setup(flutterPluginBinding)
    }
    
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        cleanup()
    }
    
    private fun setup(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        eventStreamHandler = ScgatewayEventsStreamHandler()
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "scgateway_events")
        eventChannel.setStreamHandler(eventStreamHandler)
        
        setupNotificationListener()
        Log.d(TAG, "SCGateway events channel setup complete")
    }
    
    private fun setupNotificationListener() {
        notificationObserver?.let { NotificationCenter.removeObserver(it) }
        
        notificationObserver = { notification ->
            try {
                val jsonString = notification.userInfo?.get(ScgNotification.STRINGIFIED_PAYLOAD_KEY) as? String
                if (jsonString != null) {
                    Log.d(TAG, "Sending notification to Flutter: $jsonString")
                    eventStreamHandler.sendEvent(jsonString)
                } else {
                    Log.w(TAG, "Notification has no JSON payload")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error processing notification: ${e.message}")
                eventStreamHandler.sendError("NOTIFICATION_PARSE_ERROR", e.message, e.toString())
            }
        }
        
        notificationObserver?.let { 
            NotificationCenter.addObserver(it)
            Log.d(TAG, "Notification observer added")
        }
    }
    
    private fun cleanup() {
        try {
            notificationObserver?.let { 
                NotificationCenter.removeObserver(it)
                Log.d(TAG, "Notification observer removed")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error removing notification observer: ${e.message}")
        }
        
        notificationObserver = null
        eventChannel.setStreamHandler(null)
        Log.d(TAG, "SCGateway events cleanup complete")
    }
}